File: windows/src/Notchle.Core/Engine/SplitMix64.cs
namespace Notchle.Core;

/// SplitMix64 (Steele, Lea, Flood 2014), bit-for-bit identical to
/// Sources/NotchleCore/Engine/SplitMix64.swift so the same seed gives the same shuffle on both
/// platforms (shared vector: /spec/shuffle-vectors.json). A mutable struct: keep it in a field,
/// never copy it.
internal struct SplitMix64(ulong seed)
{
    private ulong _state = seed;

    public ulong Next()
    {
        unchecked
        {
            _state += 0x9E3779B97F4A7C15UL;
            var z = _state;
            z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9UL;
            z = (z ^ (z >> 27)) * 0x94D049BB133111EBUL;
            return z ^ (z >> 31);
        }
    }

    /// Fisher-Yates with our own index draw, exactly as the Swift version consumes numbers.
    public T[] Shuffled<T>(IEnumerable<T> items)
    {
        var array = items.ToArray();
        for (var i = array.Length - 1; i > 0; i--)
        {
            var j = (int)(Next() % (ulong)(i + 1));
            (array[i], array[j]) = (array[j], array[i]);
        }
        return array;
    }
}

File: Sources/NotchleCore/Engine/SplitMix64.swift
import Foundation

/// SplitMix64 (Steele, Lea, Flood 2014). Tiny, fast and, unlike `SystemRandomNumberGenerator`,
/// reproducible from a seed on every platform, which the engine needs for deterministic shuffles.
struct SplitMix64: RandomNumberGenerator, Sendable, Hashable {
    private var state: UInt64

    init(seed: UInt64) {
        state = seed
    }

    mutating func next() -> UInt64 {
        state &+= 0x9E37_79B9_7F4A_7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58_476D_1CE4_E5B9
        z = (z ^ (z >> 27)) &* 0x94D0_49BB_1331_11EB
        return z ^ (z >> 31)
    }

    /// Fisher-Yates with our own index draw, so the order does not depend on how the standard
    /// library's `shuffle(using:)` happens to consume random numbers in a given Swift version.
    mutating func shuffle<T>(_ array: inout [T]) {
        guard array.count > 1 else { return }
        for i in stride(from: array.count - 1, to: 0, by: -1) {
            let j = Int(next() % UInt64(i + 1))
            array.swapAt(i, j)
        }
    }

    mutating func shuffled<T>(_ array: [T]) -> [T] {
        var copy = array
        shuffle(&copy)
        return copy
    }
}
